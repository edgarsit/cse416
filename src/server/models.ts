import {
  getDiscriminatorModelForClass, getModelForClass,
} from '@typegoose/typegoose';
import {
  User, GPD, Student, Description, Fields, ScrapedCourse, ScrapedCourseSet,
} from '../common/model';

export const UserModel = getModelForClass(User);
export const GPDModel = getDiscriminatorModelForClass(UserModel, GPD);
export const StudentModel = getDiscriminatorModelForClass(UserModel, Student);

export const ScrapedCourseSetModel = getModelForClass(ScrapedCourseSet);
export const ScrapedCourseModel = getModelForClass(ScrapedCourse);
const cmp: { [k: string]: string | undefined } = {
  '=': '$eq', '>': '$gt', '<': '$lt', '!=': '$neq',
};

function entries<T extends { [key: string]: any }, K extends keyof T>(o: T): [keyof T, T[K]][] {
  return Object.entries(o) as any;
}

/** get the fields from the discription from the query string */
export function getQS<T>(originalURL: string, d: Description<Fields<T>>): { [P in keyof T]?: any } {
  const params = new URL(originalURL, 'http://localhost').searchParams;
  const it = entries(d).map(
    ([k, ve]) => {
      const { ty } = ve;
      const v = params.get(k);
      if (v == null || v === '') {
        return null;
      }
      if (ty === String) {
        return [k, { $regex: v }];
      } if (ty === Number) {
        const c = cmp[`${params.get(`${k}_cmp`)}`];
        if (!c) {
          return null;
        }
        return [k, { [c]: v }];
      } if (ty === Boolean) {
        const i = ve.map!.findIndex((x) => x === v);
        if (i === -1) {
          return null;
        }
        return [k, i];
      }
      throw new Error();
    },
  ).filter(<U>(x: U): x is NonNullable<U> => x != null);
  return Object.fromEntries(it);
}

export function copyStudentWithPermissions(
  body: { [k in keyof Student]: string }, user: User,
): Partial<Student> {
  if (user.__t !== 'GPD') {
    // TODO field by field check
    throw Error('Permission denied');
  }

  const it = entries(Student.fields).map(
    ([k, ve]) => {
      if (body[k] != null) {
        return [k, body[k]];
      }
      console.warn(`missing field ${k}`);
      return null;
    },
  ).filter(<U>(x: U): x is NonNullable<U> => x != null);
  return Object.fromEntries(it);
}
