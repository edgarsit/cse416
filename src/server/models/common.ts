import { prop } from '@typegoose/typegoose';
import { WhatIsIt } from '@typegoose/typegoose/lib/internal/constants';
import { BasePropOptions } from '@typegoose/typegoose/lib/types';

export function rprop(options?: BasePropOptions, kind?: WhatIsIt): PropertyDecorator {
    if (options !== undefined) { options.required = true; }
    return prop(options, kind);
}