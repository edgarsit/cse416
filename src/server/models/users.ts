import { DocumentType, getModelForClass, getDiscriminatorModelForClass } from '@typegoose/typegoose';
import { rprop } from './common';
import { DegreeRequirements, CoursePlan } from './courses';



export class User {
  @rprop({ unique: true })
  public userName!: string;

  @rprop()
  public password!: string;
}

export const UserModel = getModelForClass(User);

export class GPD extends User {
  public async importCourseOfferings(this: DocumentType<GPD>, file: any) {
    // TODO
  }
}

export const GPDModel = getDiscriminatorModelForClass(UserModel, GPD);

export class Student extends User {
  @rprop()
  public department!: string

  @rprop()
  public track!: string

  @rprop()
  public requirementVersion!: string

  @rprop()
  public gradSemester!: string

  @rprop()
  public coursePlan!: string

  @rprop()
  public graduated!: boolean

  @rprop()
  public comments!: boolean

  @rprop()
  public sbu_id!: number

  @rprop()
  public degreeRequirements!: DegreeRequirements

  public async shareCoursePlan(this: DocumentType<Student>, c: CoursePlan) {
    // TODO
  }
}

export const StudentModel = getDiscriminatorModelForClass(UserModel, Student);
