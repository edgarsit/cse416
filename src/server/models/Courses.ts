import { Ref } from '@typegoose/typegoose';

import { rprop } from './common';
import { Student } from './users';

export class DegreeRequirements {
    @rprop()
    public department!: string

    @rprop()
    public track!: string

    @rprop()
    public requirementVersion!: string

    @rprop()
    public creditsNeededToGraduate!: number

    @rprop({ type: () => [String] })
    public requirements!: [string]
}

export class CourePlanComment {
    @rprop({ ref: () => Student })
    public author!: Ref<Student>

    @rprop()
    public text!: string
}

export class CoursePlan {
    @rprop({ type: () => [[String]] })
    public comments!: [string][]

    @rprop({ type: () => [Course] })
    public courses!: Course[]
}

export class Course {
    @rprop()
    public department!: string

    @rprop()
    public courseNum!: string

    @rprop()
    public section!: string

    @rprop()
    public year!: Date

    @rprop()
    public timeslot!: String

    @rprop({ ref: () => Course })
    public prerequisites!: Ref<Course>[]
}
