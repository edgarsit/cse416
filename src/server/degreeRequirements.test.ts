/**
 * @jest-environment node
 */

import { mongoose } from '@typegoose/typegoose';
import fs, { promises as fsp } from 'fs';
import path from 'path';
import { DegreeRequirementsModel } from './models';

const dataDir = 'data/';
const files = fs.readdirSync(dataDir).filter((x) => x.endsWith('json'));

const removeId = (o) => {
  const s = JSON.stringify(o, (k, v) => {
    if (k === '_id') {
      return undefined;
    }
    return v;
  });
  JSON.parse(s);
};

describe('insert', () => {
  beforeAll(async () => {
    await mongoose.connect('mongodb://localhost:27017/', {
      useNewUrlParser: true,
      useUnifiedTopology: true,
      useFindAndModify: false,
      useCreateIndex: true,
      dbName: 'cse416',
    });
  });

  afterAll(async () => {
    await mongoose.disconnect();
  });

  for (const f of files) {
    const filepath = path.resolve(dataDir, f);
    // eslint-disable-next-line no-loop-func
    it(`degree-req-import ${f}`, async () => {
      const dr = JSON.parse(await fsp.readFile(filepath, { encoding: 'utf8' }));
      const m = await DegreeRequirementsModel.create(dr);
      const drm = await DegreeRequirementsModel.findById(m._id);
      expect(removeId(drm)).toMatchSnapshot();
      await DegreeRequirementsModel.findByIdAndRemove(m._id);
    });
  }
});
