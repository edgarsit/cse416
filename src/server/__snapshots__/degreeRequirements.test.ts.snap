// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`insert degree-req-import AMS Degree Requirements1.json 1`] = `undefined`;

exports[`insert degree-req-import AMS Degree Requirements2.json 1`] = `undefined`;

exports[`insert degree-req-import AMS Degree Requirements3.json 1`] = `undefined`;

exports[`insert degree-req-import BMI Degree Requirements1.json 1`] = `undefined`;

exports[`insert degree-req-import BMI Degree Requirements2.json 1`] = `undefined`;

exports[`insert degree-req-import BMI Degree Requirements3.json 1`] = `undefined`;

exports[`insert degree-req-import CSE Degree Requirements1.json 1`] = `undefined`;

exports[`insert degree-req-import CSE Degree Requirements2.json 1`] = `undefined`;

exports[`insert degree-req-import CSE Degree Requirements3.json 1`] = `undefined`;

exports[`insert degree-req-import ECE Degree Requirements1.json 1`] = `undefined`;

exports[`insert degree-req-import ECE Degree Requirements2.json 1`] = `undefined`;

exports[`insert degree-req-import ECE Degree Requirements3.json 1`] = `undefined`;
